### PR TITLE
Fixing #1295 convert UnsafeCell to RefCell in CloneableService

### LINKED ISSUE #1295: Undefined behavior arrising from use of `UnsafeCell` in CloneableService
Futures created by the `HttpServiceHandler` (indirectly) contain references to shared `UnsafeCell` containing the service. With some creative use of thread local storage, it is possible for the poll function of one future to call the poll function of the other. This leads to two simultaneous mutable references into the service, which can cause undefined behavior.

I have written a demonstration of this below. 

```
use std::task::{Context, Poll};
use std::pin::Pin;
use std::future::Future;
use actix_service::{fn_factory, ServiceFactory, Service};
use actix_http::{HttpService, Request, Response, Protocol};
use futures::future::{ready, Ready};
use tokio::io::{AsyncRead, AsyncWrite};
use std::cell::Cell;
use actix_http::error::DispatchError;
use actix_http::http::{StatusCode, Error};
use futures::task::noop_waker;

thread_local! {
    static FUTURE_TO_POLL: Cell<Option<Pin<Box<dyn Future<Output = Result<(), DispatchError>>>>>> = Cell::new(None);
}

struct DangerousService(Option<Box<u8>>);

impl Service for DangerousService {
    type Request = Request;
    type Response = Response;
    type Error = Error;
    type Future = Ready<Result<Response, Error>>;

    fn poll_ready(&mut self, ctx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        unimplemented!();
    }

    fn call(&mut self, req: Self::Request) -> Self::Future {
        if let Some(mut future_to_poll) = FUTURE_TO_POLL.with(|cell| cell.take()) {
            if let Some(reference) = self.0.as_ref() {
                let waker = noop_waker();
                let mut context = Context::from_waker(&waker);
                let _ = future_to_poll.as_mut().poll(&mut context);

                println!("This read will fail {}", *reference);
            }
            ready(Ok(Response::new(StatusCode::OK)))
        } else {
            self.0 = None;
            ready(Ok(Response::new(StatusCode::OK)))
        }
    }
}

struct FakeAsync(bool);

impl AsyncRead for FakeAsync {
    fn poll_read(mut self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<Result<usize, std::io::Error>> {
        if self.0 {
            self.0 = false;
            let to_write = b"GET /hello HTTP/1.0\n\n";
            buf[0..to_write.len()].copy_from_slice(to_write);
            Poll::Ready(Ok(to_write.len()))
        } else {
            Poll::Pending
        }
    }
}

impl AsyncWrite for FakeAsync {
    fn poll_write(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<Result<usize, std::io::Error>> {
        Poll::Ready(Ok(buf.len()))
    }

    fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), std::io::Error>> {
        Poll::Ready(Ok(()))
    }

    fn poll_shutdown(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), std::io::Error>> {
        Poll::Ready(Ok(()))
    }
}

#[actix_rt::main]
async fn main() {
    let service_factory = fn_factory(|| {
        ready(Ok::<_, Error>(DangerousService(Some(Box::new(5)))))
    });
    let http_service: HttpService<FakeAsync, _, _> = HttpService::new(service_factory);
    let mut handler = http_service.new_service(()).await.unwrap();
    let future1 = handler.call((FakeAsync(true), Protocol::Http1, None));
    let future2 = handler.call((FakeAsync(true), Protocol::Http1, None));

    FUTURE_TO_POLL.with(move |cell| cell.set(Some(Box::pin(future1))));
    future2.await.unwrap();
}
```

### PR DESCRIPTION
**Pull request by dunnock**
This is fixing #1295 . With benchmarks in https://github.com/actix/actix-web/pull/1298 it can be seen that it does not have noticeable impact on performance.

This fix introduces that instead of UB it will lead to service thread panic. Service thread will be restarted thus service capacity will be reduced for some time. 

Should this be mentioned anywhere in open docs since CloneableService is not visible there. If so what is the best place?

Fixes #1295

### CODE REVIEW

**Review by robjtede — APPROVED** [review]
Checked against example code provided by @sebzim4500 and confirmed correct panic on BorrowMutError @ ./actix-http/src/cloneable.rs:38

**Comment by dunnock** [comment]
@robjtede just in case if it's waiting for my merge - I do not have write access to do so.

**Review by JohnTitor — APPROVED** [review]

**Comment by JohnTitor** [comment]
Thanks!

**Pull request merged**
